### ISSUE TITLE
Refactor Callbacks and Event monitors to be more generic

### ISSUE THREAD

**Issue by Deblintrake09** [wazuh/wazuh-qa#4159]
| Target version | Related Epic   |
|--------------------|-----------------------|
|           4.5         |  #4157             |

## Description
This Issue aims to make sure the `event_checker` functions and `callbacks` are made as generic as possible, while avoiding duplicated or too specific code.

### Event Checker
Regarding the `make_callback` and `check_event` functions the ones to be used are the ones located in the `wazuh_testing/wazuh_testing/event_monitor.py`

https://github.com/wazuh/wazuh-qa/blob/a474f439ba7b60f0e0e39524e7d0572607ac7599/deps/wazuh_testing/wazuh_testing/event_monitor.py#L6-L43

The duplicated functions in each `modules/.../event_monitor.py`  file should be replaced by the more generic version and when used, the correct prefixes must be passed in.

### Callbacks

When extracting callbacks work must be done to make them as generic as possible and remove duplicated code. For this, if possible, of every callback the matching line should be used and then used on the monitor (If the `check_event` is not used) by feeding the regex into the `generate_monitoring_callback` function and then apply any further processing in the callback.

- Examples:

The following callback:

https://github.com/wazuh/wazuh-qa/blob/a474f439ba7b60f0e0e39524e7d0572607ac7599/deps/wazuh_testing/wazuh_testing/fim.py#L1166-L1170

Could be refactored as:

```
CB_AUDIT_RULE_REMOVE = r'.* Audit rule removed.'

def callback_audit_removed_rule(line):
    match = generate_monitoring_callback(CB_AUDIT_RULE_REMOVE)
```

This way callbacks can be stored together and be changed in one place every time this message is needed.

## Considerations
- It is important to check when refactoring the callbacks if a callback has been already extracted to the corresponding `modules` folder. That version is considered to be the most recent in this case.
- If a callback is to be used in only one test module, creating callback functions should be avoided, unless the callback is complex or is known to be used in other tests that have not been already migrated.

**Comment by QU3B1M** [comment]
### First refactor proposal 

This is the first proposed generic `EventMonitor`that could be consumed on any module by only passing the monitored file and regex pattern to it. This class is created using the monitor created by Jonathan on his [qa-framework](https://github.com/wazuh/wazuh-qa-framework/)

> This option will be evaluated with QA and Core teams

```Python
class EventMonitor:
    """Class to monitor a file and check if the content matches with the specified callback.

    Attributes:
        monitored_file (str): File path to monitor.
        pattern (str): Pattern that will be evaluated for each log line.
        timeout (int): Max time to monitor and trigger the callback.
        accumulations (int): Number of expected times to match with the callback.
        only_new_events (boolean): True for only checking new lines, False to take into account all file lines.
        error_message (str): Error message to show if the timeout exception is raised.
        callback_result (*): It will store the result returned by the callback call if it is not None.
    """

    def __init__(self, monitored_file, pattern, timeout=10, accumulations=1, only_new_events=False,
                 error_message=None):
        """
        Initialize the EventMonitor class.

        Args:
            monitored_file: File path to monitor.
            pattern: Pattern that will be evaluated for each log line.
            timeout: Max time to monitor and trigger the callback.
            accumulations: Number of expected times to match with the callback.
            only_new_events: True for only checking new lines, False to take into account all file lines.
            error_message: Error message to show if the timeout exception is raised.
        """
        self.monitored_file = monitored_file
        self.pattern = pattern
        self.timeout = timeout
        self.accumulations = accumulations
        self.only_new_events = only_new_events
        self.error_message = error_message
        self.callback_result = None

        self.__validate_parameters()
        self.__start()

    def __validate_parameters(self):
        """Validate if the specified file can be monitored."""
        # Check that the monitored file exists
        if not os.path.exists(self.monitored_file):
            raise ValueError(f"File {self.monitored_file} does not exist")

        # Check that the monitored file is a file
        if not os.path.isfile(self.monitored_file):
            raise TypeError(f"{self.monitored_file} is not a file")

        # Check that the program can read the content of the file
        if not os.access(self.monitored_file, os.R_OK):
            raise PermissionError(f"{self.monitored_file} is not readable")

    def __start(self):
        """Start the file regex monitoring"""
        matches = 0
        encoding = get_file_encoding(self.monitored_file)
        # Check if current file content lines triggers the callback (only when new events has False value)
        if not self.only_new_events:
            with open(self.monitored_file, encoding=encoding) as _file:
                for line in _file:
                    matches += self.__match_line(line)
                    if matches >= self.accumulations:
                        return

        # Start count to set the timeout
        start_time = time.time()

        # Start the file regex monitoring from the last line
        with open(self.monitored_file, encoding=encoding) as _file:
            # Go to the end of the file
            _file.seek(0, 2)
            while True:
                current_position = _file.tell()
                line = _file.readline()
                # If we have not new changes wait for the next try
                if not line:
                    _file.seek(current_position)
                    time.sleep(0.1)
                # If we have a new line, check if it matches with the callback
                else:
                    matches += self.__match_line(line)
                    # If it has triggered the callback the expected times, break and leave the loop
                    if matches >= self.accumulations:
                        break

                # Add the time processing time
                elapsed_time = time.time() - start_time

                # Raise timeout error if we have passed the timeout
                if elapsed_time > self.timeout:
                    raise TimeoutError(f"Events from {self.monitored_file} did not match with the callback" if
                                       self.error_message is None else self.error_message)

    def __match_line(self, line: str) -> bool:
        """Determine if a given line matches the current pattern using the callback function.

        Args:
            line (str): The line to match.

        Returns:
            bool: True if the line matches the pattern, False otherwise.

        """
        callback_result = re.match(self.pattern, line)

        # Update callback result only if it's not None (i.e., there was a match)
        if callback_result is not None:
            self.callback_result = callback_result

        return bool(callback_result)

```

**Comment by QU3B1M** [comment]
## Update 18/05/2023 - Important:exclamation: 

After the weekly meeting with the core team, we all agree that this refactor is a good starting point for the migration of the _Integration Tests_ and its _dependencies_ to the `wazuh/wazuh` repo.
So a PR was made on `wazuh/wazuh` with the first base migration, with some deps an only one test:
- wazuh/wazuh#17155

**Comment by QU3B1M** [comment]
## Ready for review :rocket: 

A new intermediate branch [`4157_tests_migration`](https://github.com/wazuh/wazuh/tree/4157_tests_migration) was made on `wazuh/wazuh` where all the migrated code will be merged in first place, including whats on the PR wazuh/wazuh#17155.

The refactored **`FileMonitor`** that replaces the `EventMonitor` is included on this PR alongside with the refactored **`callback_generator`**
- wazuh/wazuh#17155

>This refactor also closes the issue [#4158](https://github.com/wazuh/wazuh-qa/issues/4158) as the new `FileMonitor` doesn't raise `Timeout` exceptions:
>- #4158
